File: Audio Inpainting/Solvers/inpaintFrame_OMP.py
import numpy as np
from scipy.sparse import csc_matrix
from typing import Dict, Any

def inpaintFrame_OMP(problemData: Dict[str, np.ndarray], param: Dict[str, Any]) -> np.ndarray:
    """
    Inpainting method based on Orthogonal Matching Pursuit (OMP).

    Args:
        problemData (dict): A dictionary containing the observed signal to be inpainted and the indices of clean samples.
            - 'x' (np.ndarray): Observed signal to be inpainted.
            - 'Imiss' (np.ndarray): Indices of clean samples.
        param (dict): A dictionary containing the dictionary matrix (optional if param.D_fun is set), a function handle 
        that generates the dictionary matrix param.D if param.D is not given, and the analysis window.

    Returns:
        np.ndarray: Estimated frame.
    """

    x = problemData['x']
    IObs = np.where(~problemData['IMiss'])[0]
    p_N = len(x)
    E2 = param['OMPerr']**2
    E2M = E2 * len(IObs)
    wa = param['wa'](param['N'])

    if 'D' not in param:
        param['D'] = param['D_fun'](param)

    Dict = param['D'][IObs, :]
    W = 1 / np.sqrt(np.diag(Dict.T @ Dict))
    Dict = Dict @ np.diag(W)
    xObs = x[IObs]

    residual = xObs
    maxNumCoef = param['sparsityDegree']
    indx = []
    currResNorm2 = E2M * 2
    j = 0
    while currResNorm2 > E2M and j < maxNumCoef:
        j += 1
        proj = Dict.T @ residual
        pos = np.argmax(np.abs(proj))
        indx.append(pos)
        a = np.linalg.pinv(Dict[:, indx[:j]]) @ xObs
        residual = xObs - Dict[:, indx[:j]] @ a
        currResNorm2 = np.sum(residual**2)

    indx = indx[:len(a)]

    Coeff = csc_matrix((param['D'].shape[1], 1))
    if len(indx) > 0:
        Coeff[indx] = a
        Coeff = W * Coeff

    y = param['D'] @ Coeff

    return y

File: AudioInpaintingToolbox/Solvers/inpaintFrame_OMP.m
function y = inpaintFrame_OMP(problemData,param)
% Inpainting method based on OMP 
%
% Usage: y = inpaintFrame_OMP(problemData,param)
%
%
% Inputs:
%          - problemData.x: observed signal to be inpainted
%          - problemData.Imiss: Indices of clean samples
%          - param.D - the dictionary matrix (optional if param.D_fun is set)
%          - param.D_fun - a function handle that generates the dictionary 
%          matrix param.D if param.D is not given. See, e.g., DCT_Dictionary.m and Gabor_Dictionary.m
%          - param.wa - Analysis window
%
% Outputs:
%          - y: estimated frame
%
%
% -------------------
%
% Audio Inpainting toolbox
% Date: June 28, 2011
% By Valentin Emiya, Amir Adler, Michael Elad, Maria Jafari
% This code is distributed under the terms of the GNU Public License version 3 (http://www.gnu.org/licenses/gpl.txt).
% ========================================================
% To do next: use a faster implementation of OMP

%% Load data and parameters

x = problemData.x;
IObs = find(~problemData.IMiss);
p.N = length(x);
   E2 = param.OMPerr^2;
   E2M=E2*length(IObs);

   wa = param.wa(param.N);


%% Build and normalized dictionary
% build the dictionary matrix if only the dictionary generation function is given
if ~isfield(param,'D')
   param.D = param.D_fun(param);
end
Dict=param.D(IObs,:);
W=1./sqrt(diag(Dict'*Dict));
Dict=Dict*diag(W);
xObs=x(IObs);

%% OMP iterations
residual=xObs;
maxNumCoef = param.sparsityDegree;
indx = [];
currResNorm2 = E2M*2; % set a value above the threshold in order to have/force at least one loop executed
j = 0;
while currResNorm2>E2M && j < maxNumCoef,
   j = j+1;
   proj=Dict'*residual;
   [dum pos] = max(abs(proj));
   indx(j)=pos;
   
   a=pinv(Dict(:,indx(1:j)))*xObs;
   
   residual=xObs-Dict(:,indx(1:j))*a;
   currResNorm2=sum(residual.^2);
end

%% Frame Reconstruction
indx(length(a)+1:end) = [];

Coeff = sparse(size(param.D,2),1);
if (~isempty(indx))
   Coeff(indx) = a;
   Coeff = W.*Coeff;
end
y = param.D*Coeff;

return
